### DISCUSSION TITLE
Optimizing API Calls

### DISCUSSION THREAD

**Discussion by StoyanDimitrov** [WeblateOrg/weblate#13032] [Q&A]
I have build myself a dashboard where I monitor translation progress of my watched projects and get information when strings need to be taken care of. When I need a look at that info I just open the dashboard it makes the API calls needed, and after few seconds I get that presented in a nice purple pseudo-table. Unfortunately now due to the large(?) amount of watched projects (being around 80 components in 20 projects) I hit some API rate limit. Most probably those are 1 minute limits because I'm always within the hour limit. After ~130 requests I get `HTTP/1.1 503 Service Temporarily Unavailable`. I keep a list of the projects and components of interest over which I iterate and make the following calls:
1. `https://hosted.weblate.org/api/components/{$projectSlug}/{$componentSlug}/` - from it I use `component.project.name` and `component.name`
2. `https://hosted.weblate.org/api/translations/{$projectSlug}/{$componentSlug}/{$language}/statistics/` - from it I get most of the interesting data the number of strings, translated percentage etc.
3. `https://hosted.weblate.org/api/components/{$projectSlug}/{$componentSlug}/lock/` - from it I get the locked state of the project.

I space out each request by 50 ms and additional 50 ms per project.

My request is if anyone with more experience with the API sees a way to optimize the number of queries without losing current set of data to share it with me.

The obvious solution is ether ditch the `lock` query altogether (and temporary make room for a few more components before hitting the same rate limit) or rise the waiting time between queries hoping to pass over the 1 minute mark.

**Comment by nijel** [comment]
https://docs.weblate.org/en/latest/api.html#get--api-components-(string-project)-(string-component)-statistics- gives you statistics for all the translations instead of fetching them individually

The lock state could be included in the component info, I've created https://github.com/WeblateOrg/weblate/pull/13035 to implement that.

**Reply by StoyanDimitrov** [reply]
If I understand correctly the proposed query will lower the number of API calls if I fetch statistics for more than one language which I do not.

**Reply by nijel** [reply]
Yes, that was my impression you do.
